### PR TITLE
option to allow calling RGeo "unsafe_" methods

### PR DESCRIPTION
**Pull request by ksouthworth**
I had some shapefiles that were throwing `RGeo::Error::InvalidGeometry` errors on polygons that seem to be valid (PostGIS accepts them in a `geometry` column and they can be imported using the `shp2pgsql` successfully and visualized on a map).

`rgeo` has some `unsafe_*` variants of methods available (like `unsafe_within?` instead of `within?`) mentioned here: 
https://github.com/rgeo/rgeo/blob/main/NEWS.md#validity-handling

This PR adds an `allow_unsafe_methods` option to the `RGeo::Shapefile::Reader#open` method that accepts a list of acceptable `unsafe_*` methods to call.  I implemented the option and support for just the `:within` method since that's all that was needed to fix the issue with my specific shapefile polygons, but it could be expanded to the other spatial methods as needed.

```
RGeo::Shapefile::Reader.open(file, allow_unsafe_methods: :within) do |shapefile|
   # ...
end
```

also supports array-style value for the option:

```
RGeo::Shapefile::Reader.open(file, allow_unsafe_methods: [:within]) do |shapefile|
   # ...
end
```

### CODE REVIEW

**Comment by keithdoggett** [comment]
Hi @ksouthworth thanks for the contribution I think this is a great feature to add. I'm curious how hard it would be to extend this for all unsafe methods out of the box and just make this something like `allow_unsafe_methods: true`. Quickly going through it seems like there's just 2 instances of `within?` and 1 of `contains?`. So instead of adding a rescue and then unsafe call we could do something like this:

```rb
within_method = @opts_[:allow_unsafe_methods] ? :unsafe_within? : :within?
if first_try_ >= 0 && part_data_[2].public_send(within_method, polygons_[first_try_].first[2])
```

or even

```rb
if first_try_ >= 0 && part_data_[2].public_send(@opts_[:allow_unsafe_methods] ? :unsafe_within? : :within?, polygons_[first_try_].first[2])
```

if we want to 1-line it.

What do you think?

**Comment by ksouthworth** [comment]
I definitely like the 1-line, I wasn't a big fan (stylistically) of my begin/rescue style, but I did it that way because I'm not familiar with the internals of those `rgeo` methods like `within` to know if directly calling `unsafe_within` instead of `within` would yield different results for valid geometry?

So I opted to always try the "correct" version of the method first (e.g. `within?`) and then only call the `unsafe_within?` as a fallback.  I wasn't sure if calling `unsafe_within?` skips some important stuff or not?  I guess I should go read the internals of the `rgeo` gem to get a firm answer...

**Comment by keithdoggett** [comment]
Definitely understandable. It's a reasonable assumption to make that they might behave differently.

Here's the source for the validity checking modifications.

https://github.com/rgeo/rgeo/blob/c70baceb7abd4803f137c73e295005c2bfd23740/lib/rgeo/impl_helper/validity_check.rb#L69-L96

It involves a bit of meta programming, but it basically takes the implementations that we write, and for the normal version, adds a validity check at the beginning and makes the `unsafe_` variation the original implementation that's found in the package.

It's a little contrived if you're unfamiliar with the codebase, but makes it easy to add new methods without worrying about adding a validity check at the beginning.

In short, no there's no difference in behavior between the 2 versions. Let me know if you need any more clarification.

**Comment by ksouthworth** [comment]
I have updated the code to use the 1-liner and renamed the option, and updated the README as well.

I suppose the option could just be named `unsafe_methods: true` now that I think about it

**Comment by keithdoggett** [comment]
@ksouthworth sorry about the delay here. I think naming it `unsafe_methods` makes sense.

Thinking about it more, there are 2 things we should work out before it's ready to be merged

1. This gem is compatible with older versions of RGeo where `unsafe_*` methods do not exist, so we should probably handle that somehow. We can check if unsafe methods exist, do a version check of RGeo, or make a new major release that is only compatible with RGeo 3.0+. @BuonOmo what are your thoughts on this?

2. We should add a test that ensure the option works. How hard would it be to generate test data that this works for? If it's really difficult we can just test that it accepts the option in the instance and properly sets the instance variable.

File: lib/rgeo/shapefile/reader.rb [hunk 1]
```diff
@@ -176,6 +176,7 @@ def self.open(path_, opts_ = {}, &block_)
       # block. You should use Reader::open instead.
 
       def initialize(path_, opts_ = {}) # :nodoc:
+        @opts = opts_
```
Comment (BuonOmo) [inline lib/rgeo/shapefile/reader.rb:179]: Since every other option is treated by handling it at that stage, I'd rather:
```suggestion
        @allow_unsafe = opts_[:allow_unsafe_methods]
```

**Comment by BuonOmo** [comment]
@keithdoggett I think I'd restrain bumping a major for this, but rather do something like this (pseudocode):

```ruby
@allow_unsafe = false
if opts_[:allow_unsafe_methods]
  if rgeo_version_too_low
    warn "ignored" 
  else
    @allow_unsafe = opts_[:allow_unsafe_methods]
  end
end
```

And remove this message in next major (we'd have to note it somewhere for sure). WDYT?

---

Thinking again, if you have time to publish a major, it's also fine by me

**Comment by keithdoggett** [comment]
@BuonOmo yeah I think that makes more sense actually. ideally we don't have a bunch of version checking but I think it's fine for this one case

**Comment by ObiWanKeoni** [comment]
Hello all! I have a vested interest in this being merged - Canadian CSDs appear to be raising `RGeo::Error::InvalidGeometry` when downloaded directly from the Census website. I can safely ignore these records so this is the perfect solution for me. Using this forked branch as the gem source works for me but it's not up to par with our security standards.

All of that said, I'm happy to contribute wherever I can. Unfortunately, I'm not super clear on the next steps forward so any guidance there would be great. Thanks in advance!

**Comment by BuonOmo** [comment]
Hi @ObiWanKeoni I'm not sure that you can push on an already existing branch from someone else. I'd be ok with another PR if that isn't possible though. 

The steps left to me are:

- read and reply to review
- add the test to ensure this option is working
- add the warning message if rgeo version is not compatible

I think that without previous knowledge or rgeo's internal this might be up to 1 working day at most depending on your seniority 🙂.
